Mathlib/Topology/Category/Stonean/Basic.lean

```diff
@@ -0,0 +1,148 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+import Mathlib.Topology.ExtremallyDisconnected
+import Mathlib.CategoryTheory.Sites.Coherent
+import Mathlib.Topology.Category.CompHaus.Projective
+import Mathlib.Topology.Category.Profinite.Basic
+/-!
+# Extremally disconnected sets
+
+This file develops some of the basic theory of extremally disconnected sets.
+
+## Overview
+
+This file defines the type `Stonean` of all extremally (note: not "extremely"!)
+disconnected compact Hausdorff spaces, and gives it the structure of a large category.
+
+The Lean implementation: a term of type `Stonean` is a pair, considering of
+a term of type `CompHaus` (i.e. a compact Hausdorff topological space) plus
+a proof that the space is extremally disconnected.
+This is equivalent to the assertion that the term is projective in `CompHaus`,
+in the sense of category theory (i.e., such that morphisms out of the object
+can be lifted along epimorphisms).
+
+This file defines the type of all Stonean spaces, gives it the
+structure of a large category, and proves some basic observations about this
+category and various functors from it.
+
+## Main definitions
+
+* `Stonean` : the category of extremally disconnected compact Hausdorff spaces.
+* `Stonean.toCompHaus` : the forgetful functor `Stonean ⥤ CompHaus` from Stonean
+  spaces to compact Hausdorff spaces
+* `Stonean.toProfinite` : the functor from Stonean spaces to profinite spaces.
+
+-/
+universe u
+
+open CategoryTheory
+
+/-- `Stonean` is the category of extremally disconnected compact Hausdorff spaces. -/
+structure Stonean where
+  /-- The underlying compact Hausdorff space of a Stonean space. -/
+  compHaus : CompHaus.{u}
+  /-- A Stonean space is extremally disconnected -/
+  [extrDisc : ExtremallyDisconnected compHaus]
+
+namespace CompHaus
+
+/-- `Projective` implies `Stonean`. -/
+@[simps!]
+def toStonean (X : CompHaus.{u}) [Projective X] :
+    Stonean where
+  compHaus := X
+  extrDisc := by
+    apply CompactT2.Projective.extremallyDisconnected
+    intro A B _ _ _ _ _ _ f g hf hg hsurj
+    have : CompactSpace (TopCat.of A) := by assumption
+    have : T2Space (TopCat.of A) := by assumption
+    have : CompactSpace (TopCat.of B) := by assumption
+    have : T2Space (TopCat.of B) := by assumption
+    let A' : CompHaus := ⟨TopCat.of A⟩
+    let B' : CompHaus := ⟨TopCat.of B⟩
+    let f' : X ⟶ B' := ⟨f, hf⟩
+    let g' : A' ⟶ B' := ⟨g,hg⟩
+    have : Epi g' := by
+      rw [CompHaus.epi_iff_surjective]
+      assumption
+    obtain ⟨h,hh⟩ := Projective.factors f' g'
+    refine ⟨h,h.2,?_⟩
+    ext t
+    apply_fun (fun e => e t) at hh
+    exact hh
```

Comment:
I think this should be factored out into a separate instance. In Lean 4 it is not a problem to have cycles in the type class hierarchy. So for `X : CompHaus` I think we should have a cycle `extrDisc -> Projective -> extrDisc`.